include/deal.II/base/symmetric_tensor.h

```diff
@@ -1215,17 +1215,15 @@ namespace internal
       {
       case 1:
         return data[0] * sdata[0];
-      case 2:
-        return (data[0] * sdata[0] +
-                data[1] * sdata[1] +
-                NumberType<Number>::value(2.0) * data[2] * sdata[2]);
       default:
         // Start with the non-diagonal part to avoid some multiplications by
         // 2.
         Number sum = data[dim] * sdata[dim];
         for (unsigned int d=dim+1; d<(dim*(dim+1)/2); ++d)
           sum += data[d] * sdata[d];
-        sum *= NumberType<Number>::value(2.0);
+        sum += sum; // sum = sum * 2.;
```

Comment:
The compiler still generates this as a usual multiply operation, so the processor has to go through the multiply pipeline even though it is simply an increment in the exponent and no change in the mantissa whatsoever. (It'd be something else in integer operations where the compiler actually can substitute the multiplication by the left shift operation which is then cheaper/as cheap as addition.) Even though a floating point addition clearly involves fewer switches in transistors than a multiplication, it's still not clear that an addition is actually faster as that depends on the microarchitecture. A processor might have more hardware for multiplications than additions (Intel processors of the generation Haswell and Broadwell), but additions might have lower latency. Furthermore, we tend to have more multiplications than additions, so on processors with balanced multiply/add units an addition is better. But my main argument is that for multiplication by two the code needs to load `2.0` from somewhere (usual the "constant" memory) whereas adding the same value to itself does not need to load anything.

Anyway, the reason I'm using these is because I saw this done in the machine code generated by the clang compiler and thought it would make life easier in general and read just as clearly as the other version.